src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,93 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Net.Sockets
+{
+    internal class IOControlKeepAlive
+    {
+        private const uint OnOff = 1u;
+        private const uint DefaultTimeMs = 7200000u;
+        private const uint DefaultIntervalMs = 1000u;
+        private static readonly ConditionalWeakTable<Socket, IOControlKeepAlive> SocketKeepAliveCwt;
+        private static readonly int UIntSize;
+
+        private uint timeMs;
+        private uint intervalMs;
+
+        static IOControlKeepAlive()
+        {
+            SocketKeepAliveCwt = new ConditionalWeakTable<Socket, IOControlKeepAlive>();
+            UIntSize = Marshal.SizeOf<uint>();
+        }
+
+        public static bool TryGet(Socket socket, SocketOptionLevel optionLevel, SocketOptionName optionName, ref int optionValue)
+        {
+            if (!IsNeeded(optionLevel, optionName))
+                return false;
+                
+            IOControlKeepAlive ioControlKeepAlive;
+            SocketKeepAliveCwt.TryGetValue(socket, out ioControlKeepAlive);
+            optionValue = optionName == SocketOptionName.TcpKeepAliveTime ?
+                ToS(ioControlKeepAlive?.timeMs ?? DefaultTimeMs) :
+                ToS(ioControlKeepAlive?.intervalMs ?? DefaultIntervalMs);
+            return true;
+        }
+
+        public static bool TrySet(Socket socket, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
+        {
+            if (!IsNeeded(optionLevel, optionName))
+                return false;
+
+            var ioControlKeepAlive = SocketKeepAliveCwt.GetOrCreateValue(socket);
+            if (optionName == SocketOptionName.TcpKeepAliveTime)
+                ioControlKeepAlive.timeMs = ToMs(optionValue);
+            else
+                ioControlKeepAlive.intervalMs = ToMs(optionValue);
+            socket.IOControl(IOControlCode.KeepAliveValues, ioControlKeepAlive.ToByteArray(), null);
+            return true;
+        }
+
+        public IOControlKeepAlive()
+        {
+            timeMs = DefaultTimeMs;
+            intervalMs = DefaultIntervalMs;
+        }
+
+        private static bool IsNeeded(SocketOptionLevel optionLevel, SocketOptionName optionName)
```

Comment:
Could we rename this to something like `RequiresIoctl`?

Reply:
The idea was IOControlKeepAlive.IsNeeded: it would make sense to have KeepAlive.RequiresIoctl but the class has the structure required by IoControl, that's why I named it IOControlKeepAlive
What do you think?

Reply:
Ok.